src/app/clusters/ecosystem-information-server/ecosystem-information-server.cpp

```diff
@@ -0,0 +1,368 @@
+
+    if (aHomeLocation.mFloorNumber.has_value())
+    {
+        homeLocationStruct.floorNumber.SetNonNull(aHomeLocation.mFloorNumber.value());
+    }
+    else
+    {
+        homeLocationStruct.floorNumber.SetNull();
+    }
+
+    if (aHomeLocation.mAreaType.has_value())
+    {
+        homeLocationStruct.areaType.SetNonNull(aHomeLocation.mAreaType.value());
+    }
+    else
+    {
+        homeLocationStruct.areaType.SetNull();
+    }
+    return homeLocationStruct;
+}
+
+} // namespace
+
+EcosystemDeviceStruct::Builder & EcosystemDeviceStruct::Builder::SetDeviceName(std::string aDeviceName,
+                                                                               uint64_t aDeviceNameLastEditEpochUs)
+{
+    VerifyOrDie(!mIsAlreadyBuilt);
+    mDeviceName                = std::move(aDeviceName);
+    mDeviceNameLastEditEpochUs = aDeviceNameLastEditEpochUs;
+    return *this;
+}
+
+EcosystemDeviceStruct::Builder & EcosystemDeviceStruct::Builder::SetBrigedEndpoint(EndpointId aBridgedEndpoint)
+{
+    VerifyOrDie(!mIsAlreadyBuilt);
+    mBridgedEndpoint = aBridgedEndpoint;
+    return *this;
+}
+
+EcosystemDeviceStruct::Builder & EcosystemDeviceStruct::Builder::SetOriginalEndpoint(EndpointId aOriginalEndpoint)
+{
+    VerifyOrDie(!mIsAlreadyBuilt);
+    mOriginalEndpoint = aOriginalEndpoint;
+    return *this;
+}
+
+EcosystemDeviceStruct::Builder & EcosystemDeviceStruct::Builder::AddDeviceType(Structs::DeviceTypeStruct::Type aDeviceType)
+{
+    VerifyOrDie(!mIsAlreadyBuilt);
+    mDeviceTypes.push_back(std::move(aDeviceType));
+    return *this;
+}
+
+EcosystemDeviceStruct::Builder & EcosystemDeviceStruct::Builder::AddUniqueLocationId(std::string aUniqueLocationId,
+                                                                                     uint64_t aUniqueLocationIdsLastEditEpochUs)
+{
+    VerifyOrDie(!mIsAlreadyBuilt);
+    mUniqueLocationIds.push_back(std::move(aUniqueLocationId));
+    mUniqueLocationIdsLastEditEpochUs = aUniqueLocationIdsLastEditEpochUs;
+    return *this;
+}
+
+std::unique_ptr<EcosystemDeviceStruct> EcosystemDeviceStruct::Builder::Build()
+{
+    bool deviceNameIsInvalid            = mDeviceName.size() > kDeviceNameMaxSize;
+    bool originalEndpointIsInvalid      = mOriginalEndpoint == kInvalidEndpointId;
+    bool deviceTypesListIsInvalid       = mDeviceTypes.empty();
+    bool uniqueLocationIdsListIsInvalid = mUniqueLocationIds.size() > kUniqueLocationIdsListMaxSize;
+    if (mIsAlreadyBuilt || deviceNameIsInvalid || originalEndpointIsInvalid || deviceTypesListIsInvalid ||
+        uniqueLocationIdsListIsInvalid)
+    {
+        return nullptr;
+    }
+
+    for (auto & locationId : mUniqueLocationIds)
+    {
+        if (locationId.size() > kUniqueLocationIdMaxSize)
+        {
+            return nullptr;
+        }
+    }
+
+    // std::make_unique does not have access to private constructor we workaround with using new
+    std::unique_ptr<EcosystemDeviceStruct> ret{ new EcosystemDeviceStruct(
```

Comment:
consider using make_unique? Or does that fail because of private constructor?

Reply:
Fails because of private constructor. There is a comment above mentioning this